### PR TITLE
types: nullish error types in callback function's parameter for `after` and `ready` method

### PR DESCRIPTION
**Pull request by nokazn**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

I think `err` type definition in the first parameter of the callback function for `after` and `ready` methods are insufficient.
The one for `after` method can be nullish and the one for `ready` method is optional.

See also [an small example for Fastify & TS project](https://github.com/nokazn/fastify-ts-minimum-sample) if needed.
Thank you.

### CODE REVIEW

File: types/instance.d.ts [hunk 1]
```diff
@@ -190,7 +190,7 @@ export interface FastifyInstance<
   listen(port: number | string, address?: string, backlog?: number): Promise<string>;
 
   ready(): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider> & PromiseLike<undefined>;
-  ready(readyListener: (err: Error) => void): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
+  ready(readyListener: (err?: Error) => void): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
```
Comment (climba03003) [inline types/instance.d.ts:193]:
```suggestion
  ready(readyListener: (err: Error | null) => void): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
```

**Review by climba03003 — CHANGES_REQUESTED** [review]
`error` should be `null` by default. If it doesn't, better to fix it inside `avvio`.
https://github.com/fastify/avvio/blob/master/boot.js#L447-L448

**Comment by nokazn** [comment]
@Eomm 

> Why 2 different approaches?

The both are depends on current implementation.
The parameter type of `ready` method seem to pass `undefined` to callback function as the first argument when no errors.👀

ex. https://github.com/nokazn/fastify-ts-minimum-sample

```ts
import Fastify from 'fastify';

const fastify = Fastify({
  logger: true,
});

fastify.register((server, options, done) => {
  // If you uncomment the line below, you can see error on `after` event.
  // done(new Error('something occurred'));
  done();
});

fastify.after((error) => {
  console.error({ error }, 'error has occurred on `after` event');
});
fastify.ready((error) => {
  console.error({ error }, 'error has occurred on `ready` event');
});

fastify.get('/', async (request, reply) => {
  reply.type('application/json').code(200);
  return { hello: 'world' };
});

fastify.listen({ port: 3000 }, (err, address) => {
  if (err) throw err;
  fastify.log.debug(`Server is now listening on ${address}`);
});
```

↓

image

(I also think the both should be nullish not optional as  @climba03003 said)

File: types/instance.d.ts (hunk 1 above)
Comment (nokazn) [reply types/instance.d.ts:193]: @climba03003 

> error should be null by default. If it doesn't, better to fix it inside avvio.
https://github.com/fastify/avvio/blob/master/boot.js#L447-L448

I agree with you. This `null` behavior is a standard way of Node.js.

I commit your suggestion, but if the difference between the current `undefined` behavior and `null` type definition has any problem, please let me know.

Also, I'll look into avvio implementation.

**Comment by nokazn** [comment]
@mcollina 

> Thanks for opening a PR! Can you please add a unit test? We use tsd for typescript.

I've added some type tests for `after` & `ready` methods in https://github.com/fastify/fastify/pull/5191/commits/8d1e75a70800955a53dd8c172f554aa685532963. Please check them and feel free to point out if something goes wrong. Thank you.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]

**Review by Uzlopak — APPROVED** [review]
LGTM

**Pull request merged**